From 9f312f744f27ffe09b5cfcd4dbd557323b3ee8ee Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Wed, 29 Apr 2020 09:53:28 +0800
Subject: [PATCH] DRV_DPM_SW-137: Re-Insertion using a dedicated CQM Deq Port
 and QoS QiD

- Display in proc Insertion qid/port/dma channel number
- Disabling this Feature with a macro since QoS Manager need to fix first
---
 drivers/net/datapath/dpm/datapath.h              |  3 +
 drivers/net/datapath/dpm/datapath_api.c          |  4 +
 drivers/net/datapath/dpm/datapath_proc.c         | 14 +++-
 drivers/net/datapath/dpm/gswip31/datapath_misc.c | 97 +++++++++++++++++++++++-
 4 files changed, 116 insertions(+), 2 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index 80bbd00bc2f4..faa192faee0d 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -95,6 +95,8 @@
 #define DP_DEBUG(flags, fmt, arg...)
 #endif				/* end of CONFIG_INTEL_DATAPATH_DBG */
 
+#define REINSERTION_NEWQMAP_ENABLE 0
+
 static inline bool is_invalid_port(unsigned int port_id)
 {
 	if (port_id >= MAX_DP_PORTS) {
@@ -775,6 +777,7 @@ extern struct dp_sched_info dp_sched_tbl[DP_MAX_INST][DP_MAX_SCHED_NUM];
 extern struct cqm_port_info dp_deq_port_tbl[DP_MAX_INST][DP_MAX_CQM_DEQ];
 extern struct bp_pmapper_dev dp_bp_dev_tbl[DP_MAX_INST][DP_MAX_BP_NUM];
 extern struct dma_chan_info *dp_dma_chan_tbl[DP_MAX_INST];
+extern u32 reinsert_deq_port;
 
 static inline struct inst_property *get_dp_port_prop(int inst)
 {
diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index 183052c234ae..f732c794b77f 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -55,6 +55,10 @@ atomic_t dp_status = ATOMIC_INIT(0);
 DP_DEFINE_LOCK(dp_lock);
 u32 dp_dbg_err = 1; /*print error */
 EXPORT_SYMBOL(dp_dbg_err);
+
+/* saving Globally, reinsert cqm deqport, to retrieve info later */
+u32 reinsert_deq_port = 0;
+
 static int dp_register_dc(int inst, u32 port_id,
 			  struct cbm_dp_alloc_complete_data *cbm_data,
 			  struct dp_dev_data *data, u32 flags);
diff --git a/drivers/net/datapath/dpm/datapath_proc.c b/drivers/net/datapath/dpm/datapath_proc.c
index 57a5392030d8..c9a0d2306fca 100644
--- a/drivers/net/datapath/dpm/datapath_proc.c
+++ b/drivers/net/datapath/dpm/datapath_proc.c
@@ -216,7 +216,7 @@ int proc_port_dump(struct seq_file *s, int pos)
 	int loop;
 	struct inst_info *info = NULL;
 	struct dma_chan_info *dma;
-	struct cqm_port_info *cqm;
+	struct cqm_port_info *cqm, *cqm_info;
 
 	if (!capable(CAP_SYS_PACCT))
 		return -1;
@@ -383,6 +383,18 @@ int proc_port_dump(struct seq_file *s, int pos)
 		   STATS_GET(port->rx_err_drop),
 		   STATS_GET(port->tx_err_drop));
 EXIT:
+#if REINSERTION_NEWQMAP_ENABLE
+	if (pos == 0 && reinsert_deq_port) {
+		cqm_info = get_dp_deqport_info(tmp_inst, reinsert_deq_port);
+		seq_printf(s, "    insertion_qid:     %d\n",
+			   cqm_info->qos_port);
+		seq_printf(s, "    insertion_deqport: %d\n", reinsert_deq_port);
+		dp_dma_parse_id(cqm_info->dma_chan, &cid, &pid, &nid);
+		seq_printf(s, "    insertion_tx_dma_ch:   0x%x(dma-ctrl=%d,port=%d,channel=%d)\n",
+			   cqm_info->dma_chan, cid, pid, nid);
+	}
+#endif
+
 	if (!seq_has_overflowed(s))
 		pos++;
 	if (pos >= MAX_DP_PORTS) {
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index 15021bee98c4..e0070391737b 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -80,6 +80,16 @@ static void init_dma_pmac_template(int portid, u32 flags)
 		dp_info->dma0_template[TEMPL_INSERT].field.redir = 0;
 		dp_info->dma0_mask_template[TEMPL_INSERT].field.redir = 0;
 
+		/* To have a specific Q for Reinsertion lookup is like
+		 * ENC=1, DEC=0, MPE1=0, MPE2=0 EP=0, EP is set as 0 to select
+		 * Lookup Mode 0
+		 */
+		dp_info->dma1_template[TEMPL_INSERT].field.enc = 1;
+		dp_info->dma1_mask_template[TEMPL_INSERT].field.enc = 0;
+		dp_info->dma1_mask_template[TEMPL_INSERT].field.dec = 0;
+		dp_info->dma1_mask_template[TEMPL_INSERT].field.mpe1 = 0;
+		dp_info->dma1_mask_template[TEMPL_INSERT].field.mpe2 = 0;
+		dp_info->dma1_mask_template[TEMPL_INSERT].field.ep = 0;
 	} else if (flags & DP_F_FAST_DSL) { /*sometimes with pmac*/
 		/* For normal single DSL upstream, there is no pmac at all*/
 		dp_info->dma1_template[TEMPL_NORMAL].field.dec = 1;
@@ -215,7 +225,7 @@ void dump_tx_dma_desc(struct dma_tx_desc_0 *desc_0,
 		pr_err("tx desc_0/1/2/3 NULL\n");
 		return;
 	}
-	pr_info(" DMA Descripotr:D0=0x%08x D1=0x%08x D2=0x%08x D3=0x%08x\n",
+	pr_info(" DMA Descriptor:D0=0x%08x D1=0x%08x D2=0x%08x D3=0x%08x\n",
 		*(u32 *)desc_0, *(u32 *)desc_1,
 		*(u32 *)desc_2, *(u32 *)desc_3);
 	pr_info("  DW0:redir=%d res=%d tunl=%d flow=%d ether=%d subif=0x%04x\n",
@@ -988,6 +998,83 @@ static void dp_platform_color_table_set(int inst)
 	dp_gswip_remark_dscp_set(GSW_MARKING_DSCP_AF, inst);
 }
 
+#if REINSERTION_NEWQMAP_ENABLE
+static int dp_reinsert_q_map(int inst, struct pmac_port_info *port_info,
+			     struct cbm_cpu_port_data *cpu_data)
+{
+	struct cbm_dp_alloc_data *re_insertion;
+	struct ppv4_q_sch_port q_port = {0};
+	struct cqm_port_info *c_info;
+	cbm_queue_map_entry_t lookup = {0};
+	int cbm_inst = dp_port_prop[inst].cbm_inst;
+
+	re_insertion = &cpu_data->re_insertion;
+
+	if (!is_deqport_valid(re_insertion->deq_port))
+		return DP_FAILURE;
+
+	q_port.cqe_deq = re_insertion->deq_port;
+	q_port.tx_pkt_credit = re_insertion->tx_pkt_credit;
+	q_port.tx_ring_addr = re_insertion->txpush_addr_qos;
+	q_port.tx_ring_size = re_insertion->tx_ring_size;
+
+	c_info = get_dp_deqport_info(inst, q_port.cqe_deq);
+
+	/*Store Ring Info */
+	c_info->tx_pkt_credit = re_insertion->tx_pkt_credit;
+	c_info->txpush_addr = (void *)re_insertion->txpush_addr;
+	c_info->txpush_addr_qos = (void *)re_insertion->txpush_addr_qos;
+	c_info->tx_ring_size = re_insertion->tx_ring_size;
+	c_info->dp_port = 0;/* CPU */
+
+	if (dp_dbg_flag & DP_DBG_FLAG_REG) {
+		pr_info("Re-Insertion Store CPU ring info\n");
+		pr_info("  ring_address[%d]=0x%p\n",
+			q_port.cqe_deq,
+			c_info->txpush_addr);
+		pr_info("  ring_address_push[%d]=0x%px\n",
+			q_port.cqe_deq,
+			c_info->txpush_addr_qos);
+		pr_info("  ring_size[%d]=%d\n",
+			q_port.cqe_deq,
+			c_info->tx_ring_size);
+		pr_info("  credit[%d]=%d\n",
+			q_port.cqe_deq,
+			c_info->tx_pkt_credit);
+	}
+
+	q_port.inst = inst;
+	q_port.dp_port = PMAC_CPU_ID;
+	if (alloc_q_to_port(&q_port, 0)) {
+		pr_err("alloc_q_to_port fail for re-insertion\n");
+		return DP_FAILURE;
+	}
+
+	c_info->dma_chan = re_insertion->dma_chan;
+	c_info->qos_port = q_port.qid;
+	c_info->q_node = q_port.q_node;
+	reinsert_deq_port = q_port.cqe_deq;
+
+	/* EP=0 ENC=1, DEC=0, MPE1=0, MPE2=0 use for Re-Insertion Q */
+	lookup.ep = 0;
+	lookup.enc = 1;
+	lookup.dec = 0;
+	lookup.mpe1 = 0;
+	lookup.mpe2 = 0;
+
+	cbm_queue_map_set(cbm_inst, q_port.qid, &lookup,
+			  CBM_QUEUE_MAP_F_FLOWID_L_DONTCARE |
+			  CBM_QUEUE_MAP_F_FLOWID_H_DONTCARE |
+			  CBM_QUEUE_MAP_F_TC_DONTCARE);
+
+	/* Alloc and Enable of Re-Insertion Port is done in CQM by default */
+	DP_DEBUG(DP_DBG_FLAG_REG, "Re-Insertion deq_port=%d Initialized to QiD %d\n",
+		 re_insertion->deq_port, q_port.qid);
+
+	return DP_SUCCESS;
+}
+#endif
+
 int dp_platform_queue_set(int inst, u32 flag)
 {
 	int ret, i;
@@ -1140,6 +1227,14 @@ int dp_platform_queue_set(int inst, u32 flag)
 			return -1;
 		}
 	}
+
+#if REINSERTION_NEWQMAP_ENABLE
+	/* Re-Insertion use a New Qid -> Deq Port 23 -> Dma1Tx Ch 16 */
+	if (dp_reinsert_q_map(inst, port_info, &cpu_data)) {
+		pr_err("DPM Fail to enable re-insertion Q\n");
+		return -1;
+	}
+#endif
 	return 0;
 }
 
